library/core/src/num/dec2flt/decimal.rs

```diff
@@ -205,29 +205,48 @@ impl Decimal {
 pub fn parse_decimal(mut s: &[u8]) -> Decimal {
     let mut d = Decimal::default();
     let start = s;
-    s = s.skip_chars(b'0');
-    parse_digits(&mut s, |digit| d.try_add_digit(digit));
-    if s.first_is(b'.') {
-        s = s.advance(1);
+
+    while let Some((&b'0', s_next)) = s.split_first() {
+        s = s_next;
+    }
+
+    while let Some((&ch, s_next)) = s.split_first() {
+        let digit = ch.wrapping_sub(b'0');
+        if digit < 10 {
+            d.try_add_digit(digit);
+            s = s_next;
+        } else {
+            break;
+        }
+    }
+
+    if let Some((b'.', s_next)) = s.split_first() {
+        s = s_next;
         let first = s;
         // Skip leading zeros.
         if d.num_digits == 0 {
-            s = s.skip_chars(b'0');
+            while let Some((&b'0', s_next)) = s.split_first() {
+                s = s_next;
+            }
         }
         while s.len() >= 8 && d.num_digits + 8 < Decimal::MAX_DIGITS {
-            // SAFETY: s is at least 8 bytes.
-            let v = unsafe { s.read_u64_unchecked() };
+            let v = s.read_u64();
             if !is_8digits(v) {
                 break;
             }
-            // SAFETY: d.num_digits + 8 is less than d.digits.len()
-            unsafe {
-                d.digits[d.num_digits..].write_u64_unchecked(v - 0x3030_3030_3030_3030);
-            }
+            d.digits[d.num_digits..].write_u64(v - 0x3030_3030_3030_3030);
             d.num_digits += 8;
-            s = s.advance(8);
+            s = &s[8..];
+        }
+        while let Some((&ch, s_next)) = s.split_first() {
+            let digit = ch.wrapping_sub(b'0');
+            if digit < 10 {
+                d.try_add_digit(digit);
+                s = s_next;
+            } else {
+                break;
+            }
         }
-        parse_digits(&mut s, |digit| d.try_add_digit(digit));
```

Comment:
parse_digits here used to be called in a couple places, now it looks like we've just duplicated/inlined that code. Similar process has happened with skip_chars, too.

Is there a reason for those changes? I'm not convinced they make sense - the general pattern of eating a particular character or digits seems like it _could_ be potentially optimized (e.g., by loading into a u32/u64 and finding the first non-zero byte with leading_zeroes) but duplicating the code will make that harder to try across the board.

I'd be happy to accept the changes which remove unsafe presuming they are performance-neutral, but some of these other changes feel less clear cut.